app/assets/javascripts/components/autocomplete.js

```diff
@@ -0,0 +1,175 @@
+      source: debounce(this.handleSearchQuery, 300).bind(this),
+      tNoResults: function () { return statusMessage }
+    })
+
+    this.enhanceSearchBox(this.$label, $submitButton, this.$searchContainer)
+  }
+
+  Autocomplete.prototype.enhanceSearchBox = function ($label, $submitButton, $searchContainer) {
+    // as we're removing the old search element we want to keep the old label and text
+    // but hide it as we're using the placeholder
+    var $labelClone = $label.cloneNode(true)
+    $labelClone.className = 'govuk-visually-hidden'
+
+    // we'll place the submit button back so that users
+    // are assured this is a search box
+    var $submitButtonClone = $submitButton.cloneNode(true)
+
+    // remove the old search box and insert the copy of the label and submit button
+    this.$module.removeChild($searchContainer)
+    this.$module.insertAdjacentElement('beforebegin', $labelClone)
+    this.$module.querySelector('.app-autocomplete-search__wrapper')
+      .insertAdjacentElement('afterend', $submitButtonClone)
+  }
+
+  Autocomplete.prototype.handleOnConfirm = function (query) {
+    // if GA available trigger click event submission on selected suggestion
+    this.trackSelectedOption(query)
+
+    // update the form input with the suggestion value the user selected
+    var $enhancedInput = this.$module.querySelector('input')
+    $enhancedInput.value = query
+
+    // dispatch the custom form change event to trigger the livesearch functionality
+    var $form = document.querySelector('.js-live-search-form')
+    var customEvent = document.createEvent('HTMLEvents')
+    customEvent.initEvent('customFormChange', true, false)
+    $form.dispatchEvent(customEvent)
+  }
+
+  Autocomplete.prototype.trackSelectedOption = function (query) {
+    // built-in onConfirm method only provides acces to the query
+    // so we need to find our own index of the selected option
+    // +1 as index start with 0
+    var $availableSuggestionsNodeArray = Array.prototype.slice.call(this.$module.querySelector('.app-autocomplete-search__menu').childNodes)
+    // get data from the node to use in tracking
+    var trackingDataOptions = {}
+    for (var i = 0; i < $availableSuggestionsNodeArray.length; i++) {
+      var node = $availableSuggestionsNodeArray[i]
+      if (node.innerText === query) {
+        trackingDataOptions.text = node.innerText
+        trackingDataOptions.position = node.getAttribute('aria-posinset')
+        trackingDataOptions.numberOfSuggestions = node.getAttribute('aria-setsize')
+        break
+      }
+    }
+
+    window.GOVUK.SearchAnalytics.trackEvent(
+      'suggestionClicked',
+      trackingDataOptions.position
+    )
+  }
+
+  Autocomplete.prototype.handleSearchQuery = function (query, populateResults) {
+    statusMessage = 'Searching...'
+    var matchingCachedResults = this.cachedSuggestions.filter(function (suggestion) {
+      return suggestion.toLowerCase().indexOf(query.toLowerCase()) !== -1
```

Comment:
If you get suggestions for a query, and then delete a bunch of characters, won't this just return the same highly-specific suggestions, rather than widening to give more generic suggestions?

I was imagining something like this:

```js
if (!(query in cache)) {
    cache[query] = fetch_suggestions(query);
    // maybe some logic to keep the cache under a certain number of queries
}
return cache[query];
```

Reply:
hmm 
- When you type `breakfast` quickly, you get 2 results. 
If you then start deleting characters, suggestions don't change until you the base word (`break`)

- When you type `break` quickly, you get 8 results. 

when characters are deleted, the query variable gets updated and is checked against the strings stored in the cache

as both phrases contain the same base word (`break` in this case) and that exists in the cache, how can we know best to trigger the fetching of new suggestions?

Reply:
I think fetching suggestions whenever that exact query isn't cached is probably fine, as the delay will prevent it from being every character (unless someone is typing slowly I guess)

Reply:
> ```js
> if (!(query in cache)) {
>     cache[query] = fetch_suggestions(query);
>     // maybe some logic to keep the cache under a certain number of queries
> }
> return cache[query];
> ```

this is essentially what we're doing here https://github.com/alphagov/finder-frontend/blob/c31e0f7c58fa34d717be3c09caaddb2380054fa2/app/assets/javascripts/components/autocomplete.js#L175
just not adding to the array, replacing it.

we could push to the array (at the start or at the end) and limit the array to N items, but I wonder if it's worth complicating the logic to ensure we have the latest items in the array.
It might not be needed, as I wonder hoe many queries will the user do before visiting a result and thus clearing the cached array?

Reply:
I don't think we need a queue - I doubt the user will attempt more than a couple phrases before clicking through, so at most 40 suggestions assuming we're hitting the max of 8 returned suggestions per phrase. No need to overcomplicate.